### PR TITLE
New CreateEdit documentation

### PR DESCRIPTION
**Pull request by fzaninotto**
- [x] Introduction
- [x] Form validation
- [x] `<Create>`
- [x] `<CreateBase>`
- [x] `<Edit>`
- [x] `<EditBase>`
- [x] `<EditGuesser>`
- [x] `useCreateContext`
- [x] `useCreateController`
- [x] `useEditContext`
- [x] `useEditController`
- [x] `<Form>`
- [x] `<SimpleForm>`
- [x] `<TabbedForm>`
- [x] `<AccordionForm>`
- [x] `<WizardForm>`
- [x] `<CreateDialog>`
- [x] `<EditDialog>`
- [x] `<Toolbar>`
- [x] `<SaveButton>`
- [x] `useSaveContext`

### CODE REVIEW

File: docs/Create.md [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+
+```jsx
+import * as React from 'react';
+import { useNotify, useRedirect, Create, SimpleForm } from 'react-admin';
+
+const PostCreate = () => {
+    const { reset } = useFormContext();
+    const notify = useNotify();
+
+    const  onSuccess: () => {
```
Comment (WiXSL) [inline docs/Create.md:467]:
```suggestion
    const  onSuccess = () => {
```

File: docs/Create.md [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+
+## Save And Add Another
+
+Whe users need to create several records in a row, a good UX is to stay on the Create form after a successfull submission, and to empty that form to allow a new entry. 
+
+Setting the `<Create redirect={false}>` prop only solves part of the problem: the form still needs to be emptied. That's why the right implementation for this use case is to use the `mutationOptions` prop:
+
+```jsx
+import * as React from 'react';
+import { useNotify, useRedirect, Create, SimpleForm } from 'react-admin';
```
Comment (WiXSL) [inline docs/Create.md:461]:
```suggestion
import { useNotify, Create, SimpleForm } from 'react-admin';
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — COMMENTED** [review]
The Show and Edit tutorials looks really good! 💪

**Review by WiXSL — CHANGES_REQUESTED** [review]
Throughout the `EditTutorial` code examples, the `Title` component is missing its import from react-admin

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/Create.md [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+    </Admin>
+);
+
+export default App;
+```
+
+You can customize the `<Create>` component using the following props:
+
+* [`actions`](#actions): override the actions toolbar with a custom component
+* [`aside`](#aside-component): component to render aside to the main content
```
Comment (WiXSL) [inline docs/Create.md:56]:
```suggestion
* [`aside`](#aside): component to render aside to the main content
```

File: docs/Create.md [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+* `children`: the components that renders the form
+* `className`: passed to the root component
+* [`component`](#component): override the root component
+* [`disableAuthentication`](#disable-authentication): disable the authentication check
```
Comment (WiXSL) [inline docs/Create.md:60]:
```suggestion
* [`disableAuthentication`](#disableauthentication): disable the authentication check
```

File: docs/Edit.md [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+You can customize the `<Edit>` component using the following props:
+
+* [`actions`](#actions): override the actions toolbar with a custom component
+* [`aside`](#aside-component): component to render aside to the main content
+* `children`: the components that renders the form
+* `className`: passed to the root component
+* [`component`](#component): override the root component
+* [`disableAuthentication`](#disable-authentication): disable the authentication check
+* [`id`](#id): the id of the record to edit
+* [`mutationMode`](#mutation-mode): switch to optimistic or pessimistic mutations (undoable by default)
```
Comment (WiXSL) [inline docs/Edit.md:70]:
```suggestion
* [`mutationMode`](#mutationmode): switch to optimistic or pessimistic mutations (undoable by default)
```

File: docs/Edit.md [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+```
+
+You can customize the `<Edit>` component using the following props:
+
+* [`actions`](#actions): override the actions toolbar with a custom component
+* [`aside`](#aside-component): component to render aside to the main content
+* `children`: the components that renders the form
+* `className`: passed to the root component
+* [`component`](#component): override the root component
+* [`disableAuthentication`](#disable-authentication): disable the authentication check
```
Comment (WiXSL) [inline docs/Edit.md:68]:
```suggestion
* [`disableAuthentication`](#disableauthentication): disable the authentication check
```

File: docs/Edit.md [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+    </Admin>
+);
+
+export default App;
+```
+
+You can customize the `<Edit>` component using the following props:
+
+* [`actions`](#actions): override the actions toolbar with a custom component
+* [`aside`](#aside-component): component to render aside to the main content
```
Comment (WiXSL) [inline docs/Edit.md:64]:
```suggestion
* [`aside`](#aside): component to render aside to the main content
```

File: docs/EditBase.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  );
+};
+```
+
+You can customize the `<EditBase>` component using the following props, documented in the `<Edit>` component:
+
+* `children`: the components that renders the form
+* [`disableAuthentication`](./Edit.md#disable-authentication): disable the authentication check
+* [`id`](./Edit.md#id): the id of the record to edit
+* [`mutationMode`](./Edit.md#mutation-mode): switch to optimistic or pessimistic mutations (undoable by default)
```
Comment (WiXSL) [inline docs/EditBase.md:47]:
```suggestion
* [`mutationMode`](./Edit.md#mutationmode): switch to optimistic or pessimistic mutations (undoable by default)
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/EditBase.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
+---
+layout: default
+title: "The EditBase Component"
+---
+
+# `<EditBase>`
+
+The `<EditBase>` component is a headless version of [`<Edit>`](./Edit.md): it fetches a record based on the URL, prepares a form submit handler, and renders its children. 
+
+If does that by calling [`useEditController`](./useEditController.md), and by putting the result in an `EditContext`.
```
Comment (WiXSL) [inline docs/EditBase.md:10]:
```suggestion
It does that by calling [`useEditController`](./useEditController.md), and by putting the result in an `EditContext`.
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/Edit.md [hunk 10]
```diff
@@ -0,0 +1,560 @@
...
+                ...
+            </SimpleForm>
+        </Edit>
+    );
+}
+```
+
+**Tip**: If you want to have different success side effects based on the button clicked by the user (e.g. if the creation form displays two submit buttons, one to "save and redirect to the list", and another to "save and display an empty form"), you can set the `mutationOptions` prop on the `<SaveButton>` component, too.
+
+Similarly, you can override the failure side effects with an `onError` option. By default, when the save action fails at the dataProvider level, react-admin shows an error notification.
```
Comment (WiXSL) [inline docs/Edit.md:343]:
```suggestion
Similarly, you can override the failure side effects with an `onError` option. By default, when the save action fails at the dataProvider level, react-admin shows a notification error.
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/Create.md [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+);
+```
+
+Note that the `redirect` prop is ignored if you set [the `mutationOptions` prop](#mutationoptions). See that prop for how to set a different redirection path in that case. 
+
+If you want to allow the user to enter several records one after the other, setting `redirect` to `false` won't make it, as the form isn't emptied by default. You'll have to empty the form using the `mutationOptions`, and this option disables the `redirect` prop. Check [the Save And Add Another section](#save-and-add-another) for more details.
+
+## `resource`
+
+Components based on `<Create>` are often used as `<Resource create>` props, and therefore rendered when the URL matches `/[resource]/create`. The `<Create>` component generates a call to `dataProvider.create()` using the ressource name from the URL by default.
```
Comment (WiXSL) [inline docs/Create.md:263]:
```suggestion
Components based on `<Create>` are often used as `<Resource create>` props, and therefore rendered when the URL matches `/[resource]/create`. The `<Create>` component generates a call to `dataProvider.create()` using the resource name from the URL by default.
```

File: docs/Create.md [hunk 12]
```diff
@@ -0,0 +1,459 @@
...
+
+You can decide to use a `<Create>` component in another path, or embedded in a page using another ressource name (e.g. in a Dialog). In that canse, you can explicitely set the `ressource` name:
```
Comment (WiXSL) [inline docs/Create.md:265]:
```suggestion
You can decide to use a `<Create>` component in another path, or embedded in a page using another resource name (e.g. in a Dialog). In that case, you can explicitly set the `resource` name:
```

File: docs/Edit.md [hunk 13]
```diff
@@ -0,0 +1,560 @@
+---
+layout: default
+title: "The Edit Component"
+---
+
+# `<Edit>`
+
+The `<Edit>` component is the main component for edition pages. It fetches a record based on the URL, preparess a form submit handler, and renders the page title and actions. It is not responsible for rendering the actual form - that's the job of its child component (usually a form component, like [`<SimpleForm>`](./SimpleForm.md)). This form component uses its children ([`<Input>`](./Inputs.md) components) to render each form input.
```
Comment (WiXSL) [inline docs/Edit.md:8]:
```suggestion
The `<Edit>` component is the main component for edition pages. It fetches a record based on the URL, prepares a form submit handler, and renders the page title and actions. It is not responsible for rendering the actual form - that's the job of its child component (usually a form component, like [`<SimpleForm>`](./SimpleForm.md)). This form component uses its children ([`<Input>`](./Inputs.md) components) to render each form input.
```

File: docs/Edit.md [hunk 14]
```diff
@@ -0,0 +1,560 @@
...
+        ...
+    </Edit>
+);
+```
+
+Note that the `redirect` prop is ignored if you set [the `mutationOptions` prop](#mutationoptions). See that prop for how to set a different redirection path in that case. 
+
+## `resource`
+
+Components based on `<Edit>` are often used as `<Resource edit>` props, and therefore rendered when the URL matches `/[resource]/[id]`. The `<Edit>` component generates a call to `dataProvider.update()` using the ressource name from the URL by default.
```
Comment (WiXSL) [inline docs/Edit.md:445]:
```suggestion
Components based on `<Edit>` are often used as `<Resource edit>` props, and therefore rendered when the URL matches `/[resource]/[id]`. The `<Edit>` component generates a call to `dataProvider.update()` using the resource name from the URL by default.
```

File: docs/Edit.md [hunk 15]
```diff
@@ -0,0 +1,560 @@
...
+
+You can decide to use a `<Edit>` component in another path, or embedded in a page using another ressource name (e.g. in a Dialog). In that canse, you can explicitely set the `ressource` name:
```
Comment (WiXSL) [inline docs/Edit.md:447]:
```suggestion
You can decide to use a `<Edit>` component in another path, or embedded in a page using another resource name (e.g. in a Dialog). In that case, you can explicitly set the `resource` name:
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/Edit.md [hunk 16]
```diff
@@ -0,0 +1,560 @@
...
+        ...
+    </Edit>
+);
+```
+
+## `id`
+
+Components based on `<Edit>` are often used as `<Resource edit>` props, and therefore rendered when the URL matches `/[resource]/[id]`. The `<Edit>` component generates a call to `dataProvider.update()` using the id from the URL by default.
+
+You can decide to use a `<Edit>` component in another path, or embedded in a page editing a related record (e.g. in a Dialog). In that canse, you can explicitely set the `id` value:
```
Comment (WiXSL) [inline docs/Edit.md:192]:
```suggestion
You can decide to use a `<Edit>` component in another path, or embedded in a page editing a related record (e.g. in a Dialog). In that case, you can explicitly set the `id` value:
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/useEditController.md [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+```
+{% endraw %}
+
+**Tip**: If you just use the return value of `useEditController` to put it in an `EditContext`, use [the `<EditBase>` component](./EditBase.md) instead for simpler markup.
+
+`useEditController` accepts an options argument, with the following fields, all optional:
+
+* [`disableAuthentication`](./Edit.md#disable-authentication): disable the authentication check
+* [`id`](./Edit.md#id): the id of the record to edit
+* [`mutationMode`](./Edit.md#mutation-mode): switch to optimistic or pessimistic mutations (undoable by default)
```
Comment (WiXSL) [inline docs/useEditController.md:51]:
```suggestion
* [`mutationMode`](./Edit.md#mutationmode): switch to optimistic or pessimistic mutations (undoable by default)
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/useEditContext.md [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+title: "The useEditContext hook"
+---
+
+# `useEditContext`
+
+Whenever react-admin displays an edition page, it creates an `EditContext` to store the record, the submit callback, and other data.
+
+The `CreateContext` is available to descendants of:
+
+- `<Create>`,
```
Comment (WiXSL) [inline docs/useEditContext.md:12]:
```suggestion
- `<Edit>`,
```

File: docs/useEditContext.md [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+- `<CreateGuesser>`,
```
Comment (WiXSL) [inline docs/useEditContext.md:13]:
```suggestion
- `<EditGuesser>`,
```

File: docs/useEditContext.md [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+- `<CreateBase>`,
```
Comment (WiXSL) [inline docs/useEditContext.md:14]:
```suggestion
- `<EditBase>`,
```

File: docs/useEditContext.md [hunk 21]
```diff
@@ -0,0 +1,64 @@
+---
+layout: default
+title: "The useEditContext hook"
+---
+
+# `useEditContext`
+
+Whenever react-admin displays an edition page, it creates an `EditContext` to store the record, the submit callback, and other data.
+
+The `CreateContext` is available to descendants of:
```
Comment (WiXSL) [inline docs/useEditContext.md:10]:
```suggestion
The `EditContext` is available to descendants of:
```

File: docs/useEditContext.md [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+
+- `<Create>`,
+- `<CreateGuesser>`,
+- `<CreateBase>`,
+
+All descendant components can therefore access the Create context, using the `useCreateContext` hook.
```
Comment (WiXSL) [inline docs/useEditContext.md:16]:
```suggestion
All descendant components can therefore access the Edit context, using the `useEditContext` hook. 
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/useCreateController.md [hunk 23]
```diff
@@ -0,0 +1,66 @@
+---
+layout: default
+title: "The useCreateController hook"
+---
+
+# `useCreateController`
+
+The `useCreateController` hook contains the logic of [the `<Create>` component](./Create.md): it prepares a form submit handler, and returns the data and callbacks necessary to render an Createion view.
```
Comment (WiXSL) [inline docs/useCreateController.md:8]:
```suggestion
The `useCreateController` hook contains the logic of [the `<Create>` component](./Create.md): it prepares a form submit handler, and returns the data and callbacks necessary to render a Creation view. 
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/useCreateContext.md [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+            <TextInput source="title" />
+            <TextInput source="views" />
+        </SimpleForm>
+    );
+};
+```
+
+## Return Value
+
+The `useCreateContext` hook returns an object with the same keys as returned by [`useCreateController`](#useCreateController):
```
Comment (WiXSL) [inline docs/useCreateContext.md:44]:
```suggestion
The `useCreateContext` hook returns an object with the same keys as returned by [`useCreateController`](./useCreateController.md):
```

File: docs/useCreateContext.md [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+---
+
+# `useCreateContext`
+
+Whenever react-admin displays a creation page, it creates a `CreateContext` to store the submit callback.
+
+The `CreateContext` is available to descendants of:
+
+- `<Create>`,
+- `<CreateGuesser>`,
```
Comment (WiXSL) [inline docs/useCreateContext.md:13]:
```suggestion
```

File: docs/useCreateController.md [hunk 26]
```diff
@@ -0,0 +1,66 @@
+---
+layout: default
+title: "The useCreateController hook"
+---
+
+# `useCreateController`
+
+The `useCreateController` hook contains the logic of [the `<Create>` component](./Create.md): it prepares a form submit handler, and returns the data and callbacks necessary to render an Createion view. 
+
+React-admin calls `useCreateController` internally when you use the `<Create>`, `<CreateBase>`, or `<CreateGuesser>` component.
```
Comment (WiXSL) [inline docs/useCreateController.md:10]:
```suggestion
React-admin calls `useCreateController` internally, when you use the `<Create>`, or `<CreateBase>` component.
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/EditTutorial.md [hunk 27]
```diff
@@ -321,6 +493,257 @@ export const BookEdit = () => (
...
 ```
 
-And that’s it! React-admin components are not magic, they are React components designed to let you focus on the business logic and avoid repetitive tasks. The original example was almost 60 lines of code, the react-admin one only a quarter of that.
+React-admin components are not magic, they are React components designed to let you focus on the business logic and avoid repetitive tasks. 
+
+Tip: Actually, `<Edit>` does more than the code it replaces in the previous example: it handles notification and redirection upon submission, it sets the page title, and handles the error logic.
+
+## Setting Default Values
+
+It's the Form and Ipunt components' responsibility to define default values.
```
Comment (WiXSL) [inline docs/EditTutorial.md:502]:
```suggestion
It's the Form and Input component's responsibility to define default values. 
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/SimpleForm.md [hunk 28]
```diff
@@ -0,0 +1,448 @@
...
+        </SimpleForm>
+    </Create>
+);
+```
+
+`<SimpleForm>` calls react-hook-form's `useForm` hook, and places the result in a `FormProvider` component. This means you can take advantage of the [`useFormContext`](https://react-hook-form.com/api/useformcontext) and [`useFormState`](https://react-hook-form.com/api/useformstate) hooks to access the form state.
+
+Here are all the props you can set on the `<SimpleForm>` component:
+
+* [`component`](#componnent)
```
Comment (WiXSL) [inline docs/SimpleForm.md:36]:
```suggestion
* [`component`](#component)
```

File: docs/SimpleForm.md [hunk 29]
```diff
@@ -0,0 +1,448 @@
...
+* [`defaultValues`](#defaultvalues)
+* [`id`](#id)
+* [`noValidate`](#novalidate)
+* [`onSubmit`](#onsubmit)
+* [`sx`](#sx-css-api)
+* [`toolbar`](#toolbar)
+* [`validate`](#validation)
```
Comment (WiXSL) [inline docs/SimpleForm.md:43]:
```suggestion
* [`validate`](#validate)
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/SimpleForm.md [hunk 30]
```diff
@@ -0,0 +1,448 @@
...
+
+# `<SimpleForm>`
+
+The `<SimpleForm>` creates a `<form>` to edit a record, and renders its children (usually Input components) in a simple layout, one child per row.
+
+![simple form](./img/simple-form.png)
+
+## Usage
+
+`<SimpleForm>` reads the `record` from the `RecordContext`, uses it to initialize the defaultValues of a `<Form>`, renders its children in a MUI `<Stack>`, and renders a toolbar with a `<SaveButton>` that calls the `save` callback prepared by the edit of the create controller when pressed.
```
Comment (WiXSL) [inline docs/SimpleForm.md:14]:
```suggestion
`<SimpleForm>` reads the `record` from the `RecordContext`, uses it to initialize the defaultValues of a `<Form>`, renders its children in a MUI `<Stack>`, and renders a toolbar with a `<SaveButton>` that calls the `save` callback prepared by the edit or the create controller when pressed. 
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/SimpleForm.md [hunk 31]
```diff
@@ -0,0 +1,448 @@
...
+* [`sx`](#sx-css-api)
+* [`toolbar`](#toolbar)
+* [`validate`](#validate)
+* [`warnWhenUnsavedChanges`](#warnwhenunsavedchanges)
+
+Additional props are passed to [the `useForm` hook](https://react-hook-form.com/api/useform).
+
+## `component`
+
+`<SimpleForm>` renders a MUI `<CardContent>` by default. You replaxce it by any component you want as wrapper, just pass it as the `component` prop.
```
Comment (WiXSL) [inline docs/SimpleForm.md:50]:
```suggestion
`<SimpleForm>` renders a MUI `<CardContent>` by default. You replace it by any component you want as wrapper, just pass it as the `component` prop.
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/SimpleForm.md [hunk 32]
```diff
@@ -0,0 +1,448 @@
...
+{% endraw %}
+
+## `toolbar`
+
+By default, `<SimpleForm>` renders a toolbar at the bottom of the form, containing:
+
+- a submit button on Creation pages,
+- a submit button and a delete button on Edition pages. 
+
+If you want to tweak the look and feel of that toolbar, add or remove buttons, pass yout own toolbar component to the form using the `toolbar` prop.
```
Comment (WiXSL) [inline docs/SimpleForm.md:171]:
```suggestion
If you want to tweak the look and feel of that toolbar, add or remove buttons, pass your own toolbar component to the form using the `toolbar` prop.
```

File: docs/SimpleForm.md [hunk 33]
```diff
@@ -0,0 +1,448 @@
...
+- One that redirects to a blank `<Create>` view after creation (allowing bulk creation)
+
+![Form toolbar](./img/form-toolbar.png)
+
+For that use case, use the `<SaveButton>` component with a custom `onSuccess` prop:
+
+{% raw %}
+```jsx
+import * as React from "react";
+import { Create, SimpleForm, SaveButton, Toolbar, useRedirect } from 'react-admin';
```
Comment (WiXSL) [inline docs/SimpleForm.md:199]:
```suggestion
import { Create, SimpleForm, SaveButton, Toolbar, useRedirect, useNotify } from 'react-admin';
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/TabbedForm.md [hunk 34]
```diff
@@ -0,0 +1,683 @@
...
+
+# `<TabbedForm>`
+
+`<TabbedForm>` creates a `<form>` to edit a record, and renders inputs grouped by tab. The tabs are set by using `<FormTab>` components. It is useful for forms with a lot of inputs, to reduce the time taken to change a subset of the fields.
+
+![tabbed form](./img/tabbed-form.gif)
+
+## Usage
+
+`<TabbedForm>` reads the `record` from the `RecordContext`, uses it to initialize the defaultValues of a `<Form>`, renders its children in a MUI `<Stack>`, and renders a toolbar with a `<SaveButton>` that calls the `save` callback prepared by the edit of the create controller when pressed.
```
Comment (WiXSL) [inline docs/TabbedForm.md:14]:
```suggestion
`<TabbedForm>` reads the `record` from the `RecordContext`, uses it to initialize the defaultValues of a `<Form>`, renders its children in a MUI `<Stack>`, and renders a toolbar with a `<SaveButton>` that calls the `save` callback prepared by the edit or the create controller when pressed. 
```

File: docs/TabbedForm.md [hunk 35]
```diff
@@ -0,0 +1,683 @@
...
+## Dynamic Tab Label
+
+`<TabbedForm>` often contain not only inputs, but also related data (e.g. the reviews of a product). Users appreviate that the label of such tabs show the actual number of related elements, to avoid clicking on a tab to reveal an empty list.
+
+![dynamic tab label](./img/FormTab-dynamic-label.png)
+
+To achieve that, create a custom commponent that renders a `<FormTab>` with a dynamic `label`:
+
+```jsx
+const ReviewsFormTab = (props: any) => {
```
Comment (WiXSL) [inline docs/TabbedForm.md:614]:
```suggestion
const ReviewsFormTab = props => {
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/CreateBase.md [hunk 36]
```diff
@@ -0,0 +1,51 @@
+---
+layout: default
+title: "The CreateBase Component"
+---
+
+# `<CreateBase>`
+
+The `<CreateBase>` component is a headless version of `<Create>`: it prepares a form submit handler, and renders its children. 
+
+If does that by calling `useCreateController`, and by putting the result in an `CreateContext`.
```
Comment (WiXSL) [inline docs/CreateBase.md:10]:
```suggestion
It does that by calling `useCreateController`, and by putting the result in an `CreateContext`.
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/EditTutorial.md [hunk 37]
```diff
@@ -266,41 +397,82 @@ export const BookEdit = () => {
...
+      </div>
+-   </EditContextProvider>
++   </EditBase>
+  );
+};
+```
+
+### `<Edit>` Renders Title, Fields, And Actions
+
+`<EditBase>` is a headless component: it renders only its children. But almost every edition view needs a wrapping `<div>`, a title, and a `<Card>`. That’s why react-admin provides [the `<Edit>` component](./Edit.md), which includes the `<EditBase>` component, a title built from the resource name, and even a “Show" button if the resource has a show component:
```
Comment (WiXSL) [inline docs/EditTutorial.md:442]:
```suggestion
`<EditBase>` is a headless component: it renders only its children. But almost every edition view needs a wrapping `<div>`, a title, and a `<Card>`. That’s why react-admin provides [the `<Edit>` component](./Edit.md), which includes the `<EditBase>` component, a title built from the resource name, and even a "Show" button if the resource has a show component:
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/Inputs.md [hunk 38]
```diff
@@ -34,6 +34,7 @@ All input components accept the following props:
 |-----------------| -------- |---------------------------| ------- |---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
 | `source`        | Required | `string`                  | -       | Name of the entity property to use for the input value                                                                                                              |
 | `label`         | Optional | `string`                  | -       | Input label. In i18n apps, the label is passed to the `translate` function. Defaults to the humanized `source` when omitted. Set `label={false}` to hide the label. |
+| `defaultValue`  | Optional | `any`                     | -       | Default value of the input. |
 | `validate`      | Optional | `Function` &#124; `array` | -       | Validation rules for the current property. See the [Validation Documentation](./CreateEdit.md#validation) for details.                                              |
```
Comment (WiXSL) [inline docs/Inputs.md:38]:
```suggestion
| `validate`      | Optional | `Function` &#124; `array` | -       | Validation rules for the current property. See the [Validation Documentation](./Validation.md#per-input-validation-built-in-field-validators) for details.                                              |
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/CreateEdit.md [hunk 39]
```diff
@@ -2221,85 +684,6 @@ If the form has several save buttons, you can also pass a custom `onSuccess` or
...
-        <SimpleForm toolbar={<PostEditToolbar />}>
-            // ...
-        </SimpleForm>
-    </Edit>
-);
-```
-
 ### Grouping Inputs
 
 Sometimes, you may want to group inputs in order to make a form more approachable. You may use a [`<TabbedForm>`](#the-tabbedform-component), an [`<AccordionForm>`](#the-accordionform-component) or you may want to roll your own layout. In this case, you might need to know the state of a group of inputs: whether it's valid or if the user has changed them (dirty/touched state).
```
Comment (WiXSL) [inline docs/CreateEdit.md:689]:
```suggestion
Sometimes, you may want to group inputs in order to make a form more approachable. You may use a [`<TabbedForm>`](./TabbedForm.md), an [`<AccordionForm>`](#the-accordionform-component) or you may want to roll your own layout. In this case, you might need to know the state of a group of inputs: whether it's valid or if the user has changed them (dirty/touched state).
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/Create.md [hunk 40]
```diff
@@ -0,0 +1,460 @@
...
+        ...
+    </Create>
+);
+```
+
+## `mutationOptions`
+
+You can customize the options you pass to react-query's `useMutation` hook, e.g. to override success or error side effects, by setting the `mutationOptions` prop. Refer to the [useMutation documentation](https://react-query.tanstack.com/reference/useMutation) in the react-query website for a list of the possible options.
+
+Let's see an example with the success side effect. By default, when the save action succeeds, react-admin shows a notification, and redirects to another page. You can override this behavior and pass custom success side effects by providing a `mutationOptions` prop with an `onSuccess` key:
```
Comment (djhi) [inline docs/Create.md:157]:
```suggestion
Let's see an example with the success side effect. By default, when the save action succeeds, react-admin shows a notification, and redirects the new record edit page. You can override this behavior and pass custom success side effects by providing a `mutationOptions` prop with an `onSuccess` key:
```

File: docs/Create.md [hunk 41]
```diff
@@ -0,0 +1,460 @@
...
+## `redirect`
+
+By default, submitting the form in the `<Create>` view redirects to the `<Edit>` view.
+
+You can customize the redirection by setting the `redirect` prop to one of the following values:
+
+- `'edit'`: redirect to the Edit view (the default)
+- `'list'`: redirect to the List view
+- `'show'`: redirect to the Show view
+- `false`: do not redirect
```
Comment (djhi) [inline docs/Create.md:247]: We also accept a function: `redirect((resource, id, data) => string)`

File: docs/Create.md [hunk 42]
```diff
@@ -0,0 +1,460 @@
...
+
+```jsx
+export const PostCreate = () => (
+    <Create title="New post">
+        ...
+    </Create>
+);
+```
+
+The `title` value can be a string or a React element.
```
Comment (djhi) [inline docs/Create.md:301]: Might be worth reminding users they can access the record that was passed by calling `useRecordContext`.

File: docs/Create.md [hunk 43]
```diff
@@ -0,0 +1,460 @@
...
+        Write a comment for that post
+    </Button>
+);
+```
+{% endraw %}
+
+Should you use the location `state` or the location `search`? The latter modifies the URL, so it's only necessary if you want to build cross-application links (e.g. from one admin to the other). In general, using the location `state` is a safe bet.
+
+And if you want to prefill the form with constant values, use the `defaultValues` prop on the Form tag.  
+
```
Comment (djhi) [inline docs/Create.md:423]: Might be worth mentioning you can use the `CloneButton` and provide its `record` prop for that

File: docs/Edit.md [hunk 44]
```diff
@@ -0,0 +1,562 @@
...
+    </Edit>
+);
+```
+{% endraw %}
+
+## `mutationOptions`
+
+`<Edit>` calls `dataProvider.update()` via react-query's `useMutation` hook. You can customize the options you pass to this hook, e.g. to override success or error side effects, by setting the `mutationOptions` prop. Refer to the [useMutation documentation](https://react-query.tanstack.com/reference/useMutation) in the react-query website for a list of the possible options.
+
+Let's see an example with the success side effect. By default, when the save action succeeds, react-admin shows a notification, and redirects to another page. You can override this behavior and pass custom success side effects by providing a `mutationOptions` prop with an `onSuccess` key:
```
Comment (djhi) [inline docs/Edit.md:271]:
```suggestion
Let's see an example with the success side effect. By default, when the save action succeeds, react-admin shows a notification, and redirects to the list page. You can override this behavior and pass custom success side effects by providing a `mutationOptions` prop with an `onSuccess` key:
```

File: docs/Edit.md [hunk 45]
```diff
@@ -0,0 +1,562 @@
...
+
+## `redirect`
+
+By default, submitting the form in the `<Edit>` view redirects to the `<List>` view.
+
+You can customize the redirection by setting the `redirect` prop to one of the following values:
+
+- `'list'`: redirect to the List view (the default)
+- `'show'`: redirect to the Show view
+- `false`: do not redirect
```
Comment (djhi) [inline docs/Edit.md:431]: We also accept a function `redirect((resource, id, data) => string | Path)`

File: docs/Form.md [hunk 46]
```diff
@@ -0,0 +1,190 @@
...
+);
+```
+
+**Tip**: You can include properties in the form `defaultValues` that are not listed as input components, like the `created_at` property in the previous example.
+
+**Tip**: React-admin also allows to define default values at the input level. See the [Setting default Values](./EditTutorial.md#setting-default-values) section.
+
+## `id`
+
+Normally, a submit button only works when placed inside a `<form>` tag. However, you can place a submit button outside of the form if the submit button `form_id` matches the form `id`.
```
Comment (djhi) [inline docs/Form.md:82]: :+1:

File: docs/SaveButton.md [hunk 47]
```diff
@@ -0,0 +1,197 @@
...
+- [`onClick`](#onClick)
+- [`sx`](#sx-css-api)
+- [`transform`](#transform)
+- [`type`](#type)
+
+Additional props (e.g. `color`, `variant`) are passed to [the underlying MUI `<Button>` component](https://mui.com/components/buttons/).
+
+## `icon`
+
+By default, `<SaveButton>` renders a disk icon. You can xcan pass another icon element:
```
Comment (djhi) [inline docs/SaveButton.md:61]:
```suggestion
By default, `<SaveButton>` renders a disk icon. You can can pass another icon element:
```

File: docs/SaveButton.md [hunk 48]
```diff
@@ -0,0 +1,197 @@
...
+
+```jsx
+const PostCreateToolbar = () => (
+    <Toolbar>
+        <SaveButton label="Save" onClick={() => alert('Saving...')} />
+    </Toolbar>
+);
+```
+
+`onClick` it doesn't *replace* the default submission handler, but is instead called before it. To override the default submission handler, wrap a `<SaveButton>` in a custom [`SaveContext`](./useSaveContext.md).
```
Comment (djhi) [inline docs/SaveButton.md:132]: Might be worth mentioning that you can call `preventDefault` to cancel the submission

**Review by djhi — CHANGES_REQUESTED** [review]
Awesome!

File: docs/SaveButton.md [hunk 49]
```diff
@@ -0,0 +1,197 @@
...
+    </Edit>
+);
+```
+
+`<SaveButton>` accepts the following props:
+
+- [`icon`](#icon)
+- [`label`](#label)
+- [`mutationOptions`](#mutationoptions)
+- [`onClick`](#onClick)
```
Comment (WiXSL) [inline docs/SaveButton.md:52]:
```suggestion
- [`onClick`](#onclick)
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/SimpleForm.md [hunk 50]
```diff
@@ -0,0 +1,470 @@
...
+    </Create>
+);
+
+const Separator = () => <Box pt="1em" />;
+```
+{% endraw %}
+
+## Displaying Inputs Based On Permissions
+
+You can leverage [the `usePermissions` hook](./usePermissions.md) to display a inputs if the user has the required permissions.
```
Comment (WiXSL) [inline docs/SimpleForm.md:451]:
```suggestion
You can leverage [the `usePermissions` hook](./usePermissions.md) to display inputs if the user has the required permissions.
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — APPROVED** [review]

File: docs/Create.md (hunk 42 above)
Comment (fzaninotto) [reply docs/Create.md:301]: in the context of the Create page, it doesn't make much sense IMHO

**Pull request merged**